### PR TITLE
Feature/reg847 with reg859

### PR DESCRIPTION
**Pull request by nigelhp**
This PR combines the work for:
* REG-847: Handle Technical Issues (Local Unit)
* REG-859: Refactor Retrieve Enterprise by Exact Row Key (Happy Path Only)

### CODE REVIEW

File: test/it/EnterpriseAcceptanceSpec.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+          aColumnWith(name = name, value = SampleEnterpriseName),
+          aColumnWith(name = postcode, value = SamplePostcode),
+          aColumnWith(name = legalStatus, value = SampleLegalStatus),
+          aColumnWith(name = employees, value = SampleNumberOfEmployees.toString),
+          aColumnWith(name = jobs, value = SampleJobs.toString))
+      ).mkString("[", ",", "]")
+    }}"""
+
+  info("As a SBR user")
+  info("I want to retrieve a enterprise unit with a specific period (by year and month) in time")
```
Comment (nigelhp) [inline test/it/EnterpriseAcceptanceSpec.scala]: __an__ enterprise unit ...

File: test/it/EnterpriseAcceptanceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  info("So that I can view the enterprise details via the api response")
+
+  feature("retrieve an existing enterprise unit") {
+    scenario("by exact Enterprise reference (ERN) and period") { wsClient =>
+      Given(s"a enterprise rowKey of $TargetErn and $TargetPeriod")
```
Comment (nigelhp) [inline test/it/EnterpriseAcceptanceSpec.scala]: Given setups the pre-conditions of the test.
When writing acceptance tests try not to think like a developer.  What do our business users care about row keys?

"Given an enterprise exists with Ern of ... and Period of ..."

File: test/it/EnterpriseAcceptanceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      stubHBaseFor(aEnterpriseUnitRequest(withErn = TargetErn, withPeriod = TargetPeriod).willReturn(
+        anOkResponse().withBody(EnterpriseUnitSingleMatchHBaseResponseBody)
+      ))
+
+      When(s"a enterprise unit with $TargetErn and $TargetPeriod is request")
```
Comment (nigelhp) [inline test/it/EnterpriseAcceptanceSpec.scala]: __an__ enterprise with ... is __requested__

File: test/it/EnterpriseAcceptanceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      val response = await(wsClient.url(s"/v1/periods/${Period.asString(TargetPeriod)}/enterprises/${TargetErn.value}").get())
+
+      Then(s"the response should match as follows and the details should resemble an expected enterprise with $TargetPeriod and $TargetErn")
```
Comment (nigelhp) [inline test/it/EnterpriseAcceptanceSpec.scala]: the details "should resemble"?  How closely should they resemble it? - if only half the values resemble it is that OK?

__"The details of the enterprise identified by ... should be returned."__

Categorical.  It has to be an exact match.

No point talking about "as follows" - the only thing that will follow in sbt ouput is the next test.  We want the Given / When / Then text to tell a story - and to be descriptive enough when printed in isolation by sbt to understand the purpose of the test.

File: test/it/EnterpriseAcceptanceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  private val EnterpriseUnitSingleMatchHBaseResponseBody =
+    s"""{"Row": ${
+      List(
+        aRowWith(key = s"${EnterpriseUnitRowKey(TargetErn, TargetPeriod)}", columns =
+          aColumnWith(name = ern, value = TargetErn.value),
+          aColumnWith(name = entref, value = SampleEnterpriseReference),
+          aColumnWith(name = name, value = SampleEnterpriseName),
+          aColumnWith(name = postcode, value = SamplePostcode),
+          aColumnWith(name = legalStatus, value = SampleLegalStatus),
+          aColumnWith(name = employees, value = SampleNumberOfEmployees.toString),
```
Comment (nigelhp) [inline test/it/EnterpriseAcceptanceSpec.scala]: We cannot have an acceptance test for every possible scenario, that way leads to incredibly slow builds.  However, nor is the acceptance test fixed in stone.  As your implementation develops and you discover more of the edge cases, see if some of them can be reflected in the test case.

Once you are happy you can support all fields, you may therefore want to have one of the optional fields as missing.  Of course, you have unit tests that prove you can handle all the fields correctly - but such a change helps the acceptance test reflect more of the "real world" case - rather than a perfect scenario where every single field is populated.

File: test/repository/hbase/enterprise/EnterpriseUnitRowMapperSpec.scala [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    "make a Enterprise with only mandatory fields" in new Fixture {
+      println(EnterpriseUnitRowMapper.fromRow(mandatoryVariables))
+      EnterpriseUnitRowMapper.fromRow(mandatoryVariables) shouldBe Some(SampleEnterpriseWithNoOptionalFields)
+    }
+
+  }
+
+  "Fails to create a Enterprise Unit RowMapper when" - {
```
Comment (nigelhp) [inline test/repository/hbase/enterprise/EnterpriseUnitRowMapperSpec.scala]: We are not creating Enterprise Unit RowMapper's ...

File: test/repository/hbase/enterprise/EnterpriseUnitRowMapperSpec.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    "a mandatory field is missing" in new Fixture {
+      val mandatoryColumnKeys: Iterable[String] = mandatoryVariables.keys
+      mandatoryColumnKeys.foreach { column =>
+        withClue(s"Missing field is $column") {
+          EnterpriseUnitRowMapper.fromRow(mandatoryVariables - column) shouldBe None
+        }
+      }
+    }
+
+    "a expected field has a mismatch with an expected type" in new Fixture {
```
Comment (nigelhp) [inline test/repository/hbase/enterprise/EnterpriseUnitRowMapperSpec.scala]: Not particularly descriptive ...
You are only dealing with ints, so can be more explicit - "an integer field cannot be successfully parsed" ...

Note that it is not just about non-numeric values.  What about "3.14159"?  Is that a valid employee count?

File: test/server/HBaseRestSpec.scala [hunk 8]
```diff
@@ -76,19 +76,19 @@ class HBaseRestSpec extends TestUtils with BeforeAndAfterEach with GuiceOneAppPe
...
-      mockEndpoint(enterpriseTable, Some(firstPeriod), id, None, body)
-      val resp = fakeRequest(s"/$version/periods/$firstPeriod/enterprises/$id")
-      val json = contentAsJson(resp)
-      val ent = json.validate[EnterpriseUnit]
-      status(resp) mustBe OK
-      contentType(resp) mustBe Some("application/json")
-      ent.isInstanceOf[JsSuccess[EnterpriseUnit]] mustBe true
-    }
-  }
+  //  "/v1/periods/:period/enterprises/:id" should {
```
Comment (nigelhp) [inline test/server/HBaseRestSpec.scala]: If we no longer need it - remove it

File: test/server/RouteSpec.scala [hunk 9]
```diff
@@ -35,15 +35,15 @@ class RouteSpec extends TestUtils {
...
-    "return 400 due to Invalid Key" in {
-      val search = fakeRequest("/v1/periods/201706/enterprises/1")
-      status(search) mustBe BAD_REQUEST
-    }
-
-    "return 400 with invalid date is not parsable" in {
-      val dateSearch = fakeRequest("/v1/periods/201777/enterprises/1244")
-      status(dateSearch) mustBe BAD_REQUEST
-    }
+    //    "return 400 due to Invalid Key" in {
```
Comment (nigelhp) [inline test/server/RouteSpec.scala]: as above - if we don't need it, remove it ...

File: test/support/WithWireMockHBase.scala [hunk 10]
```diff
@@ -5,19 +5,37 @@ import com.github.tomakehurst.wiremock.client.{ MappingBuilder, ResponseDefiniti
...
+import repository.hbase.unit.enterprise.EnterpriseUnitRowKey
+
 trait WithWireMockHBase extends WithWireMock with BasicAuthentication with HBaseResponseFixture { this: Suite =>
   override val wireMockPort = 8075
 
+  private val ColumnFamily = "d"
+  private val Namespace = "sbr_control_db"
+
+  private def createUrlAndThenGetHbaseJson(tableName: String, rowKey: String) =
+    getHbaseJson(
```
Comment (nigelhp) [inline test/support/WithWireMockHBase.scala]: While it doesn't comply with usual camel-case conventions, I have been trying to use "HBase" as that is how the project itself capitalises the name.

File: test/support/WithWireMockHBase.scala [hunk 11]
```diff
@@ -5,19 +5,37 @@ import com.github.tomakehurst.wiremock.client.{ MappingBuilder, ResponseDefiniti
...
+    createUrlAndThenGetHbaseJson(tableName, rowKey)
+  }
+
+  def aEnterpriseUnitRequest(withErn: Ern, withPeriod: Period): MappingBuilder = {
+    val rowKey = EnterpriseUnitRowKey(withErn, withPeriod)
+    val tableName = "enterprise"
+    createUrlAndThenGetHbaseJson(tableName, rowKey)
   }
 
   def getHbaseJson(url: String, auth: Authorization): MappingBuilder =
```
Comment (nigelhp) [inline test/support/WithWireMockHBase.scala]: Looks like this one slipped through the net (I was originally using camel-case, but eventually relented).  We should change it for consistency ...

File: test/support/sample/SampleEnterpriseUnit.scala [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+  val SampleEnterpriseId = Ern("1000000012")
+  val SampleNumberOfEmployees = 100
+  val SampleJobs = 15
+  val SampleEnterpriseReference = "someEnterpriseRef"
+  val SampleEnterpriseName = "Company Name Plc"
+  val SamplePostcode = "NP0 XXX"
+  val SampleLegalStatus = "some-LegalUnit"
+
+  val DefaultDeliminator = "~"
```
Comment (nigelhp) [inline test/support/sample/SampleEnterpriseUnit.scala]: Why is this needed here?
We only want one definition of the rowKey delimiter - and that is in the HBase object.
